Спринт 4

DenisVolc opened the pull request.

Commit:
Добавил комент в pom и добавил папку .idea/ в гитигнор

Commit:
Начал писать первый тест из проекта

Commit:
Дописла первые два теста без параметризации

Commit:
Пофиксил аасерт в orderPageTest и подтвержжение заказа в OrderPage.java

Commit:
заменить локаторы By на WebElemеnt

Commit:
Добавил 3 и 4 тесты

Commit:
Начал делать 6ой тест.

Commit:
Доделал 6ой тест, сделал 5ый тест, сделал параметризацию

Commit:
Исправил ошибки, привел код в соответсвие с ТЗ

DenisVolc commented:
Первая попытка сдать проект 4го спринта.

pomegranate282922 commented on src/test/java/Tests/FaqParmTest.java:
```diff
@@ -0,0 +1,85 @@
+import java.util.concurrent.TimeUnit;
+@RunWith(Parameterized.class)
+public class FaqParmTest {
+    private final String mainPageUrl= "https://qa-scooter.praktikum-services.ru/";
```
Строковые константы обычно делаем static final 
Можно улучшить

pomegranate282922 commented on src/test/java/Tests/FaqParmTest.java:
```diff
@@ -0,0 +1,85 @@
+    @Parameterized.Parameters // добавили аннотацию
+    public static Object[][] orderTestDataSet() {
+        return new Object[][]{
+                {0,"chrome"},
```
Браузер в параметры теста не вставляем.
Нужно исправить.

Если хочется, можно сделать выбор браузера через системную переменную.
Можно улучшить

pomegranate282922 commented on src/test/java/Tests/FaqParmTest.java:
```diff
@@ -0,0 +1,85 @@
+    }
+
+    @Parameterized.Parameters // добавили аннотацию
+    public static Object[][] orderTestDataSet() {
```
Название метода нелогичное 
Нужно исправить

pomegranate282922 commented on src/test/java/Tests/FaqParmTest.java:
```diff
@@ -0,0 +1,85 @@
+        mnPage.findFAQ();
+//      Мне показалось, что проще перебрать элементы аккордиона циклом,
+//      потому что не нужно знать заранее кол-во элементов.
+//      Да и не нужно руками забивать тестовые данные. Если бы их было не 7 а хотя бы 30, было бы грустно.
```
Нет, так нехорошо, так как тесты становятся зависимыми.

pomegranate282922 commented on src/test/java/Tests/LogoTests.java:
```diff
@@ -0,0 +1,87 @@
+    @Parameterized.Parameters // добавили аннотацию
+    public static Object[][] orderTestDataSet() {
+        return new Object[][]{
+                {"chrome"},
```
Это нужно убрать 
Нужно исправить

pomegranate282922 commented on src/test/java/Tests/LogoTests.java:
```diff
@@ -0,0 +1,87 @@
+            driver = new FirefoxDriver();
+        }
+        driver.manage().timeouts().implicitlyWait(3, TimeUnit.SECONDS);
+
```
Общие начала и концы лучше вынести в общий класс предок 
Можно улучшить

DenisVolc replied:
Сходу не получилось сделать. Если это важно поправлю в следующей итерации.

pomegranate282922 replied:
Это на улучшение.

pomegranate282922 commented on src/test/java/Tests/OrderParmTest.java:
```diff
@@ -0,0 +1,101 @@
+        this.isUpperorderButton = orderButton;
+    }
+    @Parameterized.Parameters // добавили аннотацию
+    public static Object[][] TestDataSet() {
```
Методы именуем со строчной буквы 
Нужно исправить

pomegranate282922 commented on src/test/java/Tests/OrderParmTest.java:
```diff
@@ -0,0 +1,101 @@
+                {"Иван","Иванов","Проспект Ленина д.1",0,"89151234578","mozila",false},
+                {"петр","васильев","ул. Арбат",100,"+79059876521","chrome",false},
+                {"петр","васильев","ул. Арбат",100,"+79059876521","mozila",true},
+        };
```
Браузер нужно убрать из параметров

pomegranate282922 commented on src/test/java/Tests/OrderParmTest.java:
```diff
@@ -0,0 +1,101 @@
+        orderPage.clickYesButton();
+        Assert.assertTrue("Не отображается подтверждение заказа",
+                orderPage.isAppearOrderConfirmed());
+    }
```
Здесь нужно немного изменить логику теста.
 Первый флоу ( верхняя кнопка) - тестируем минимум двумя тестовыми наборами до появления надписи Заказ оформлен.
 Второй флоу ( нижняя кнопка) - тестируем до появления формы ввода параметров заказа.
 Нужно исправить

pomegranate282922 commented on src/test/java/Tests/OrderTests.java:
```diff
@@ -0,0 +1,106 @@
+package Tests;
```
Пакеты именуем строчными буквами 
Нужно исправить

pomegranate282922 commented on src/test/java/Tests/OrderTests.java:
```diff
@@ -0,0 +1,106 @@
+
+    }
+    @Parameterized.Parameters // добавили аннотацию
+    public static Object[][] TestDataSet() {
```
Нейминг нужно исправить

pomegranate282922 commented on src/test/java/Tests/OrderTests.java:
```diff
@@ -0,0 +1,106 @@
+            Assert.assertTrue(orderPage.isAppearMetroStationError());
+            Assert.assertTrue(orderPage.isAppearPhoneNumberError());
+            Assert.assertTrue(orderPage.isAppearAddressError());
+            //На второй странце заказа нет сообщений об ошибках и не понятно какие у них локаторы,
```
Так тесты зависимы . Если упадет первая проверка, остальные не проверятся. 
нужно исправить

pomegranate282922 commented on src/test/java/Tests/OrderTests.java:
```diff
@@ -0,0 +1,106 @@
+            mainPage.clickTrackButton();//Кликнуть "Статус заказа"
+            new WebDriverWait(driver, Duration.ofSeconds(5))
+                    .until(ExpectedConditions.elementToBeClickable(mainPage.getGoTrackButton()));
+//            mainPage.setInputTrack("952951");// ввести правильный номер заказа для отладки; Почему-то в хроме этот тест будет проходить
```
А где ввод номера заказа? Не поняла.
Нужно исправить

DenisVolc replied:
Пустой номер заказа, это тоже неправильный номер заказа, поэтому я решил пропустить этот шаг. Исправлю.

pomegranate282922 replied:
Пустой это другой класс эквивалентности относительно неправильного.

pomegranate282922 commented on src/test/java/page_objects/MainPage.java:
```diff
@@ -0,0 +1,96 @@
+package page_objects;
```
Пакеты именуем без _
Нужно исправить

pomegranate282922 commented on src/test/java/page_objects/MainPage.java:
```diff
@@ -0,0 +1,96 @@
+package page_objects;
+
+import org.openqa.selenium.By;
```
Обычно если проект полностью посвящен автотестам, все кроме тестов размещают в main.
Можно улучшить

pomegranate282922 commented on src/test/java/Tests/OrderTests.java:
```diff
@@ -0,0 +1,106 @@
+
+    }
+
+
```
Отлично что выполнено доп задание!

pomegranate282922 commented on src/test/java/Tests/OrderTests.java:
```diff
@@ -0,0 +1,106 @@
+        }
+
+        @Test
+        public void wrongOrderTestN6() {
```
N6 и тп из имени лучше убрать, а дать методам говорящие логичные ясные имена.
Можно улучшить

pomegranate282922 commented on src/test/java/Tests/FaqParmTest.java:
```diff
@@ -0,0 +1,85 @@
+//        }
+//        Реализация по ТЗ:
+        mnPage.pressQuestion(questionNumber);
+        Assert.assertTrue(mnPage.isEnabledAnswer(questionNumber));
```
Нужен проверить, что все 8 текстов ответов соответствуют ожидаемым. 
нужно исправить

pomegranate282922 reviewed (commented):
Работа хорошая! Здорово, что сделано доп задание. 
замечания:
1. Нейминг методов, пакетов 
2. Логика в тестах на создание заказа 
3. Поправить тесты на faq 
4. Убрать браузер из параметров 
5. Сделать все тесты независимыми 
6. Остальное по замечаниям по коду

Commit:
Исправил замечания по первой версии проекта

pomegranate282922 commented on src/main/java/pageobjects/MainPage.java:
```diff
@@ -0,0 +1,116 @@
+    private WebElement goTrackButton;
+    @FindBy(xpath = ".//input[@class='Input_Input__1iN_Z Header_Input__xIoUq']")
+    private WebElement inputTrack;
+    private final String[]expectedText = {
```
? Что это?

DenisVolc replied:
private WebElement goTrackButton; // Это кнопка GO для проверки номера заказа
private WebElement inputTrack; // Это поле ввода номера заказа для проверки. В новом коммите переименовал эту переменную в inputTrackNumber
private final String[] expectedText = { //Это массив с ожидаемым текстом ответов на вопросы

pomegranate282922 replied:
Ожидаемые значения не являются частью Page Object. Их лучше вынести как константы в отдельное место.
Можно улучшить

pomegranate282922 commented on src/main/java/pageobjects/OrderPage.java:
```diff
@@ -0,0 +1,119 @@
+    public void clickYesButton(){
+        confirmYesButton.click();
+    }
+    public boolean isAppearOrderConfirmed(){
```
Нужно проверить сам текст, а не просто непустоту 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/FaqParmTest.java:
```diff
@@ -0,0 +1,71 @@
+import java.util.concurrent.TimeUnit;
+@RunWith(Parameterized.class)
+public class FaqParmTest {
+    static final String mainPageUrl= "https://qa-scooter.praktikum-services.ru/";
```
Статичные константы именуем заглавными буквами через нижнее подчеркивание 
Нужно исправить

pomegranate282922 commented on src/test/java/tests/LowerOrderButtonTest.java:
```diff
@@ -0,0 +1,38 @@
+
+public class LowerOrderButtonTest {
+    WebDriver driver;
+    static final String mainPageUrl= "https://qa-scooter.praktikum-services.ru/";
```
Не надо везде одну и ту же константу вводить. Достаточно одной.
Можно улучшить

pomegranate282922 reviewed (commented):
Многое исправил - супер!
Но есть еще замечания:
1. Ошибка компиляции 
2. Нейминг статичных констант 
3. Добавить проверку на текст в тесте на создание заказа 
4. Остальное по мелочи

DenisVolc commented:
> 1. Ошибка компиляции

Если речь об этих ошибках, то наставник сказал их игнорировать

SLF4J: No SLF4J providers were found.
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See https://www.slf4j.org/codes.html#noProviders for further details.

DenisVolc committed:
Вынес URL константой в отдельный класс.
Добавил проверку на текст в тесте на создание заказа.
Поправил skipTests в pom.xml

pomegranate282922 commented on src/test/java/tests/ParmOrderTest.java:
```diff
@@ -0,0 +1,92 @@
+//        проверить всплывающее окно
+        orderPage.clickYesButton();
+        Assert.assertTrue("Не отображается подтверждение заказа",
+                orderPage.isAppearOrderConfirmed("Заказ оформлен"));
```
Этот тест должен быть зеленым в ff, например. 
нужно исправить

DenisVolc replied:
В ff, он зеленый, но в хроме кнопка подтверждения заказа не работает. Я даже руками проверял. 
Я не понимаю что нужно исправить.

В первой версии проекта, тесты выполнялись и в мозиле и в хроме, но заданию надо было оставить только хром. Даже тогда в мозиле этот тест проходил, а в хроме нет.

pomegranate282922 reviewed (commented):
Все ок, только в тесте на создание заказа в проверке закралась ошибочка.

Commit:
Исправил ошибку в isAppearOrderConfirmed

Commit:
Перенес массив с ответами на вопросы expectedText в класс с константами.
Рефактор.

DenisVolc committed:
Create README.md

pomegranate282922 commented on src/main/java/constants/Constants.java:
```diff
@@ -0,0 +1,17 @@
+
+public class Constants {
+    public static final String MAIN_PAGE_URL = "https://qa-scooter.praktikum-services.ru/";
+    public static final String[] EXPECTED_TEXT = { //Это массив с ожидаемым текстом ответов на вопросы
```
Лучше более конкретное имя дать

pomegranate282922 reviewed (approved):
Ок, зачет!
